src/Uno.UI.Composition/Composition/BorderShapeVisual.skia.cs

```diff
@@ -0,0 +1,186 @@
+			{
+				_backgroundClip = Compositor.CreateRectangleClip(
+					innerArea.Left, innerArea.Top, innerArea.Right, innerArea.Bottom,
+					fullCornerRadius.Inner.TopLeft, fullCornerRadius.Inner.TopRight, fullCornerRadius.Inner.BottomRight, fullCornerRadius.Inner.BottomLeft);
+			}
+			else
+			{
+				_backgroundClip = Compositor.CreateRectangleClip(
+					outerArea.Left, outerArea.Top, outerArea.Right, outerArea.Bottom,
+					fullCornerRadius.Outer.TopLeft, fullCornerRadius.Outer.TopRight, fullCornerRadius.Outer.BottomRight, fullCornerRadius.Outer.BottomLeft);
+			}
+		}
+		else
+		{
+			CornerRadiusClip = null;
+		}
+	}
+
+	private void UpdateBackground(bool useInnerBorderBoundsAsAreaForBackground, SKRect innerArea, SKRect outerArea, SKPoint[] outerRadii, SKPoint[] innerRadii)
+	{
+		var backgroundGeometry = (CompositionPathGeometry)(BackgroundShape.Geometry ??= Compositor.CreatePathGeometry());
+		backgroundGeometry.Path ??= new CompositionPath(new SkiaGeometrySource2D());
+		var backgroundPathGeometry = ((SkiaGeometrySource2D)backgroundGeometry.Path.GeometrySource).Geometry;
+		backgroundPathGeometry.Reset();
+		var roundRect = new SKRoundRect();
+		SetRectRadii(
+			roundRect,
+			useInnerBorderBoundsAsAreaForBackground ? innerArea : outerArea,
+			useInnerBorderBoundsAsAreaForBackground ? innerRadii : outerRadii);
+		backgroundPathGeometry.AddRoundRect(roundRect);
+		backgroundPathGeometry.Close();
+	}
+
+	private void UpdateBorderShape(SKRect innerArea, SKRect outerArea, SKPoint[] outerRadii, SKPoint[] innerRadii)
+	{
+		var borderGeometry = (CompositionPathGeometry)(BorderShape.Geometry ??= Compositor.CreatePathGeometry());
+		borderGeometry.Path ??= new CompositionPath(new SkiaGeometrySource2D(new SKPath()));
+		var borderGeometrySource = (SkiaGeometrySource2D)borderGeometry.Path.GeometrySource;
+		var borderPathGeometry = borderGeometrySource.Geometry;
+
+		borderPathGeometry.Reset();
+
+		// It's important to set this every time, since borderPathGeometry.Reset will reset it.
+		borderGeometrySource.Geometry.FillType = SKPathFillType.EvenOdd;
+
+		// The order here (outer then inner) is important because of the SKPathFillType.
+		{
+			var outerRect = new SKRoundRect();
+			SetRectRadii(outerRect, outerArea, outerRadii);
+			borderPathGeometry.AddRoundRect(outerRect);
+			borderPathGeometry.Close();
+		}
+		{
+			var innerRect = new SKRoundRect();
+			SetRectRadii(innerRect, innerArea, innerRadii);
+			borderPathGeometry.AddRoundRect(innerRect);
+			borderPathGeometry.Close();
+		}
+	}
+
+	// We use this instead of the equivalent SKRoundRect.SetRectRadii because it forces the array to have a
+	// length of _exactly_ 4. We rent the SKPoint array, so we're not guaranteed to get the exact length we need.
+	private unsafe static void SetRectRadii(SKRoundRect roundRect, SKRect area, SKPoint[] radii)
+	{
+		fixed (SKPoint* radii1 = radii)
+		{
+			sk_rrect_set_rect_radii(roundRect.Handle, &area, radii1);
```

Comment:
If we are going this approach to pass a pointer, I think we could take this step further and use `stackalloc SKPoint[4]` instead of array pool.